xlr/sdk/src/validator.ts

```diff
@@ -62,31 +75,98 @@ export class XLRValidator {
         validationIssues.push(error);
       }
     } else if (xlrNode.type === "or") {
-      // eslint-disable-next-line no-restricted-syntax
+      const potentialTypeErrors: Array<{
+        type: NodeType;
+        errors: Array<ValidationMessage>;
+      }> = [];
+
       for (const potentialType of xlrNode.or) {
         const potentialErrors = this.validateType(rootNode, potentialType);
+
         if (potentialErrors.length === 0) {
           return validationIssues;
         }
+
+        potentialTypeErrors.push({
+          type: potentialType,
+          errors: potentialErrors,
+        });
       }
 
       let message: string;
+      const expectedTypes = xlrNode.or.map(
+        (node) => node.name ?? node.title ?? node.type
+      );
 
       if (xlrNode.name) {
-        message = `Does not match any of the expected types for type: '${xlrNode.name}'`;
+        message = `Received unexpected types for type: '${xlrNode.name}'\n`;
       } else if (xlrNode.title) {
-        message = `Does not match any of the expected types for property: '${xlrNode.title}'`;
+        message = `Received unexpected types for property: '${xlrNode.title}'\n`;
       } else {
-        message = `Does not match any of the types ${xlrNode.or
-          .map((node) => node.name ?? node.title ?? "<unnamed type>")
-          .join(" | ")}`;
+        message = `Does not match any of the types: ${expectedTypes.join(
+          " | "
+        )}\n`;
+      }
+
+      const nestedTypes = new Set<string>();
+
+      potentialTypeErrors.forEach((typeError) => {
+        // Exclude so nested types don't try to resolve ExpressionRef | BindingRef
+        if (typeError.type.type !== "template") {
+          typeError.errors.forEach((error) => {
+            if (error.expected) {
+              // Split by separate types if union
+              String(error.expected)
+                .split(" | ")
+                .forEach((val) => nestedTypes.add(val.trim()));
+            }
+          });
+        }
+      });
+
+      const maxValidShown = 20; // Max number of nested types to display
```

Comment:
Nitpick but do we need to redeclare this every time or can we reference it from a static constant. I think we could also change some of the below logic to just do it in one pass.